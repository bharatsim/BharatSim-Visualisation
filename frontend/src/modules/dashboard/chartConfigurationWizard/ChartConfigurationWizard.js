import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Drawer from '@material-ui/core/Drawer';
import Box from '@material-ui/core/Box';
import { makeStyles } from '@material-ui/core/styles';
import ChartConfigurationHeader from './ChartConfigurationHeader';
import ChartSelectorStep from './ChartSelectorStep';
import ChartConfigSelectorStep from './ChartConfigSelectorStep';

const useDrawerStyles = makeStyles((theme) => ({
  paper: {
    width: '45vw',
    borderRadius: theme.spacing(1),
    marginTop: theme.spacing(16),
    height: `calc(100vh - ${theme.spacing(16)}px)`,
  },
}));

function ChartConfigurationWizard({ chart: chartToEdit, isOpen, closeModal, onApply }) {
  const drawerClasses = useDrawerStyles();
  const initialStep = chartToEdit ? 1 : 0;
  const initialChartType = chartToEdit ? chartToEdit.chartType : '';
  const [activeStep, setActiveStep] = useState(initialStep);
  const [selectedChart, setSelectedChart] = useState(initialChartType);

  const existingConfig = chartToEdit ? chartToEdit.config : {};
  const chartId = chartToEdit ? chartToEdit.layout.i : undefined;

  function goToNextStep() {
    setActiveStep((prevActiveStep) => prevActiveStep + 1);
  }

  function goToChartSelectorStep() {
    setActiveStep(0);
  }

  function onClickOfChartSelectorNext(chart) {
    setSelectedChart(chart);
    goToNextStep();
  }
  function onClickOfConfigSelectorNext(config) {
    onApply(chartId, selectedChart, config);
  }

  return (
    <Drawer anchor="right" open={isOpen} onClose={closeModal} classes={drawerClasses}>
      <Box m={6}>
        <ChartConfigurationHeader
          closeModal={closeModal}
          chart={selectedChart}
          activeStep={activeStep}
        />
        <Box>
          {activeStep === 0 && <ChartSelectorStep onNext={onClickOfChartSelectorNext} />}
          {activeStep === 1 && (
            <ChartConfigSelectorStep
              chartType={selectedChart}
              existingConfig={existingConfig}
              onApply={onClickOfConfigSelectorNext}
              backToChartType={goToChartSelectorStep}
            />
          )}
        </Box>
      </Box>
    </Drawer>
  );
}

ChartConfigurationWizard.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  chart: PropTypes.shape({
    layout: PropTypes.shape({ i: PropTypes.string.isRequired }).isRequired,
    config: PropTypes.shape({}).isRequired,
    chartType: PropTypes.string.isRequired,
  }),
  closeModal: PropTypes.func.isRequired,
  onApply: PropTypes.func.isRequired,
};

ChartConfigurationWizard.defaultProps = {
  chart: null,
};

export default ChartConfigurationWizard;
