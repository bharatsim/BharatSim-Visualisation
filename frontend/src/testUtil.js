// eslint-disable-next-line import/no-extraneous-dependencies
import { fireEvent, within } from '@testing-library/react';

export function selectDropDownOption(container, dropDownId, optionId) {
  const dropDown = container.getByTestId(dropDownId);
  fireEvent.mouseDown(within(dropDown).getByRole('button'));
  const options = within(
    within(document.getElementById(`menu-${dropDownId}`)).getByRole('listbox'),
  );
  fireEvent.click(options.getByTestId(`${dropDownId}-${optionId}`));
}
