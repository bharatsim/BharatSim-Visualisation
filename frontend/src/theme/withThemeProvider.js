import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme(props) {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent {...props} />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
