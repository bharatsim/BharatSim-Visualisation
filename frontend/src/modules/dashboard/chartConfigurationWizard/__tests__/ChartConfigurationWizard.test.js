import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import { within } from '@testing-library/dom';
import ChartConfigurationWizard from '../ChartConfigurationWizard';
import withThemeProvider from '../../../../theme/withThemeProvider';
import { selectDropDownOption, withProjectLayout, withRouter } from '../../../../testUtil';

jest.mock('../../../../utils/api', () => ({
  api: {
    getDatasources: jest.fn().mockResolvedValue({
      dataSources: [
        { name: 'datasource1', _id: 'id1' },
        { name: 'datasource2', _id: 'id2' },
      ],
    }),
    getCsvHeaders: jest.fn().mockResolvedValue({
      headers: [
        { name: 'column1', type: 'number' },
        { name: 'column2', type: 'number' },
      ],
    }),
  },
}));

const ComponentWithProvider = withThemeProvider(
  withRouter(withProjectLayout(ChartConfigurationWizard)),
);

describe('Chart configuration wizard', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should match snapshot in edit mode', async () => {
    const existingChart = {
      chartType: 'lineChart',
      config: {
        chartName: 'line chart with config',
        dataSource: 'id1',
        xAxis: 'column1',
        yAxis: [{ name: 'column2' }],
      },
      layout: { i: 'widget-3', x: 6, y: null, w: 6, h: 2 },
    };
    const { findByText, getByLabelText, getByTestId } = render(
      <ComponentWithProvider
        chart={existingChart}
        closeModal={jest.fn()}
        isOpen
        onApply={jest.fn()}
      />,
    );

    await findByText('Data Source');
    expect(getByLabelText('Add chart name')).toHaveValue('line chart with config');

    await findByText('select x axis');
    const xAxisDropdown = within(getByTestId('x-axis-dropdown'));
    const yAxisDropdown = within(getByTestId('y-axis-dropdown-0'));

    expect(xAxisDropdown.getByRole('button')).toHaveTextContent('column1');
    expect(yAxisDropdown.getByRole('button')).toHaveTextContent('column2');
  });

  it('should select chart and go to next step of configure chart', async () => {
    const { findByText, getByText, getByTestId } = render(
      <ComponentWithProvider closeModal={jest.fn()} isOpen onApply={jest.fn()} />,
    );

    const lineChartOption = getByTestId('lineChart');
    fireEvent.click(lineChartOption);

    await findByText('Data Source');

    expect(getByText('Data Source')).toBeInTheDocument();
  });

  it('should call on apply function on click of on apply button of config step', async () => {
    const mockOnApply = jest.fn();
    const renderedComponent = render(
      <ComponentWithProvider closeModal={jest.fn()} isOpen onApply={mockOnApply} />,
    );
    const { findByText, getByText, getByTestId, getByLabelText } = renderedComponent;

    const lineChartOption = getByTestId('lineChart');

    fireEvent.click(lineChartOption);
    await findByText('Data Source');

    const chartNameInput = getByLabelText('Add chart name');
    fireEvent.input(chartNameInput, {
      target: { value: 'chart name' },
    });

    await selectDropDownOption(renderedComponent, 'dropdown-dataSources', 'datasource2');
    await findByText('select x axis');
    await selectDropDownOption(renderedComponent, 'x-axis-dropdown', 'column1');
    await selectDropDownOption(renderedComponent, 'y-axis-dropdown-0', 'column2');

    const applyButton = getByText('Apply');

    await act(async () => {
      expect(applyButton).not.toBeDisabled();
    });

    await act(async () => {
      fireEvent.click(applyButton);
    });

    expect(mockOnApply).toHaveBeenCalledWith(undefined, 'lineChart', {
      chartName: 'chart name',
      dataSource: 'id2',
      xAxis: 'column1',
      yAxis: [{ name: 'column2' }],
    });
  });

  it('should call on apply function on click of on apply button of config step with default chart name', async () => {
    const onApplyMock = jest.fn();
    const renderedComponent = render(
      <ComponentWithProvider closeModal={jest.fn()} isOpen onApply={onApplyMock} />,
    );
    const { findByText, getByText, getByTestId } = renderedComponent;

    const lineChartOption = getByTestId('lineChart');

    fireEvent.click(lineChartOption);
    await findByText('Data Source');

    await selectDropDownOption(renderedComponent, 'dropdown-dataSources', 'datasource2');
    await findByText('select x axis');
    await selectDropDownOption(renderedComponent, 'x-axis-dropdown', 'column1');
    await selectDropDownOption(renderedComponent, 'y-axis-dropdown-0', 'column2');

    const applyButton = getByText('Apply');

    await act(async () => {
      expect(applyButton).not.toBeDisabled();
    });

    await act(async () => {
      fireEvent.click(applyButton);
    });

    expect(onApplyMock).toHaveBeenCalledWith(undefined, 'lineChart', {
      chartName: 'Untitled Chart',
      dataSource: 'id2',
      xAxis: 'column1',
      yAxis: [{ name: 'column2' }],
    });
  });
  it('should disable apply button if any config is not selected', async () => {
    const onApplyMock = jest.fn();
    const renderedComponent = render(
      <ComponentWithProvider closeModal={jest.fn()} isOpen onApply={onApplyMock} />,
    );
    const { findByText, getByText, getByTestId } = renderedComponent;

    const lineChartOption = getByTestId('lineChart');

    fireEvent.click(lineChartOption);
    await findByText('Data Source');

    await selectDropDownOption(renderedComponent, 'dropdown-dataSources', 'datasource2');
    await findByText('select x axis');
    await selectDropDownOption(renderedComponent, 'x-axis-dropdown', 'column1');

    const applyButton = getByText('Apply').closest('button');

    await act(async () => {
      expect(applyButton).toBeDisabled();
    });
  });

  it('should get back to chart selector step on click of back to chart type', async () => {
    const renderedComponent = render(
      <ComponentWithProvider closeModal={jest.fn()} isOpen onApply={jest.fn()} />,
    );
    const { findByText, getByText, getByTestId } = renderedComponent;

    const lineChartOption = getByTestId('lineChart');

    fireEvent.click(lineChartOption);
    await findByText('Data Source');

    const backToChartTypeButton = getByText('Back to chart type').closest('button');
    fireEvent.click(backToChartTypeButton);

    expect(getByText('Choose a chart type to start configuring your chart')).toBeInTheDocument();
  });

  it('should close modal on click of close icon', () => {
    const closeModal = jest.fn();
    const { getByTestId } = render(
      <ComponentWithProvider closeModal={closeModal} isOpen onApply={jest.fn()} />,
    );

    const closeIcon = getByTestId('close-wizard');

    fireEvent.click(closeIcon);

    expect(closeModal).toHaveBeenCalled();
  });
});
