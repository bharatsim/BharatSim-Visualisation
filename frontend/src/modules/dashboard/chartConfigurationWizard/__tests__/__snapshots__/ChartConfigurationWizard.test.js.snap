// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart configuration wizard should match snapshot 1`] = `
<div
  class="MuiDrawer-root MuiDrawer-modal"
  role="presentation"
  style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiPaper-root MuiDrawer-paper makeStyles-paper-1 MuiDrawer-paperAnchorRight MuiPaper-elevation16"
    tabindex="-1"
  >
    <div
      class="MuiBox-root MuiBox-root-2"
    >
      <div
        class="MuiBox-root MuiBox-root-6"
      >
        <div
          class="MuiBox-root MuiBox-root-7"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6"
          >
            Chart Configuration Wizard
          </h6>
          
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root WithStyles(ForwardRef(IconButton))-root-8"
          data-testid="close-wizard"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root MuiBox-root-9"
      >
        <div
          class="MuiBox-root MuiBox-root-12"
        >
          <hr
            class="MuiDivider-root"
          />
          <div
            class="MuiBox-root MuiBox-root-13"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle1"
            >
              Choose a chart type to start configuring your chart
            </h6>
            <div
              class="MuiBox-root MuiBox-root-14 makeStyles-chartSelectorContainer-10"
            >
              <div
                class="MuiBox-root MuiBox-root-15"
              >
                <div
                  class="MuiBox-root MuiBox-root-16"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2"
                  >
                    Line Chart
                  </h6>
                </div>
                <div
                  class="MuiBox-root MuiBox-root-17 makeStyles-chartTypeContainer-11"
                >
                  <div
                    class="MuiBox-root MuiBox-root-21 makeStyles-buttonContainer-18"
                    data-testid="lineChart"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-22 makeStyles-imageContainer-19"
                    >
                      <img
                        alt="lineChart"
                        src="lineChart.svg"
                      />
                    </div>
                    <span
                      class="MuiTypography-root MuiTypography-caption"
                    >
                      Line Chart
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-23"
              >
                <div
                  class="MuiBox-root MuiBox-root-24"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2"
                  >
                    Bar Chart
                  </h6>
                </div>
                <div
                  class="MuiBox-root MuiBox-root-25 makeStyles-chartTypeContainer-11"
                >
                  <div
                    class="MuiBox-root MuiBox-root-26 makeStyles-buttonContainer-18"
                    data-testid="barChart"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-27 makeStyles-imageContainer-19"
                    >
                      <img
                        alt="barChart"
                        src="barChart.svg"
                      />
                    </div>
                    <span
                      class="MuiTypography-root MuiTypography-caption"
                    >
                      Bar Chart
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-28"
              >
                <div
                  class="MuiBox-root MuiBox-root-29"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2"
                  >
                    Geo Chart
                  </h6>
                </div>
                <div
                  class="MuiBox-root MuiBox-root-30 makeStyles-chartTypeContainer-11"
                >
                  <div
                    class="MuiBox-root MuiBox-root-31 makeStyles-buttonContainer-18"
                    data-testid="heatMap"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-32 makeStyles-imageContainer-19"
                    >
                      <img
                        alt="heatMap"
                        src="heatmap.svg"
                      />
                    </div>
                    <span
                      class="MuiTypography-root MuiTypography-caption"
                    >
                      Heatmap
                    </span>
                  </div>
                  <div
                    class="MuiBox-root MuiBox-root-33 makeStyles-buttonContainer-18"
                    data-testid="choroplethMap"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-34 makeStyles-imageContainer-19"
                    >
                      <img
                        alt="choroplethMap"
                        src="heatmap.svg"
                      />
                    </div>
                    <span
                      class="MuiTypography-root MuiTypography-caption"
                    >
                      Choropleth
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;
